File: src-tauri/src/commands/matcher.rs
//! Matcher Commands
//!
//! Tauri commands for channel matching operations.
//! Business logic is delegated to `services::matcher`.

use serde::Serialize;
use tauri::{AppHandle, Emitter, State};

use crate::db::models::{ChannelMapping, XmltvChannelSettings, XtreamChannel};
use crate::db::DbConnection;
use crate::matcher::{ChangedStream, M3uMatchResult, MatchStats, ProviderChanges};
use crate::services::matcher as svc;
use crate::services::matcher::MatchProgress;

/// Response type for match operations
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MatchResponse {
    pub success: bool,
    pub matched_count: usize,
    pub unmatched_count: usize,
    pub total_xmltv: usize,
    pub total_source_channels: usize,
    pub duration_ms: u64,
    pub message: String,
}

/// Response type for M3U auto-match operations
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct M3uAutoMatchResponse {
    pub success: bool,
    pub matched_count: usize,
    pub unmatched_count: usize,
    pub total_m3u_channels: usize,
    pub total_xmltv_channels: usize,
    pub duration_ms: u64,
    pub mappings_created: i32,
    pub message: String,
}

/// Run the channel matching algorithm.
#[tauri::command]
pub async fn run_channel_matching(
    app: AppHandle,
    db: State<'_, DbConnection>,
    threshold: Option<f64>,
) -> Result<MatchResponse, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    let result = svc::run_channel_matching(&mut conn, threshold, |progress| {
        match &progress {
            MatchProgress::Starting { message } => {
                let _ = app.emit("match_progress", serde_json::json!({
                    "status": "starting", "message": message
                }));
            }
            MatchProgress::Saving { message } => {
                let _ = app.emit("match_progress", serde_json::json!({
                    "status": "saving", "message": message
                }));
            }
            MatchProgress::Complete { matched, unmatched } => {
                let _ = app.emit("match_progress", serde_json::json!({
                    "status": "complete", "matched": matched, "unmatched": unmatched
                }));
            }
        }
    })?;

    Ok(MatchResponse {
        success: true,
        matched_count: result.matched_count,
        unmatched_count: result.unmatched_count,
        total_xmltv: result.total_xmltv,
        total_source_channels: result.total_source_channels,
        duration_ms: result.duration_ms,
        message: format!(
            "Matched {} of {} XMLTV channels ({} with multiple matches). {} mappings saved.",
            result.matched_count,
            result.total_xmltv,
            result.multiple_matches,
            result.mappings_saved
        ),
    })
}

/// Get current match statistics from the database.
#[tauri::command]
pub fn get_match_stats(db: State<DbConnection>) -> Result<MatchStats, String> {
    let pool = db.clone_pool();
    crate::matcher::calculate_match_stats(&pool)
        .map_err(|e| format!("Failed to calculate match stats: {}", e))
}

/// Get channel mappings for a specific XMLTV channel.
#[tauri::command]
pub fn get_channel_mappings_for_xmltv(
    db: State<DbConnection>,
    xmltv_channel_id: i32,
) -> Result<Vec<ChannelMapping>, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::get_channel_mappings_for_xmltv(&mut conn, xmltv_channel_id)
}

/// Get XMLTV channel settings.
#[tauri::command]
pub fn get_xmltv_channel_settings(
    db: State<DbConnection>,
    xmltv_channel_id: i32,
) -> Result<Option<XmltvChannelSettings>, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::get_xmltv_channel_settings(&mut conn, xmltv_channel_id)
}

/// Get the current matching threshold.
#[tauri::command]
pub fn get_match_threshold(db: State<DbConnection>) -> Result<f64, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::get_match_threshold(&mut conn)
}

/// Set the matching threshold.
#[tauri::command]
pub fn set_match_threshold(db: State<DbConnection>, threshold: f64) -> Result<(), String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::set_match_threshold(&mut conn, threshold)
}

/// Normalize a channel name (exposed for testing/debugging).
#[tauri::command]
pub fn normalize_channel_name(name: String) -> String {
    svc::normalize_channel_name(&name)
}

/// Calculate match score between two channel names (exposed for testing/debugging).
#[tauri::command]
pub fn calculate_match_score(
    xmltv_name: String,
    xtream_name: String,
    epg_id_match: bool,
    exact_name_match: bool,
) -> f64 {
    svc::calculate_match_score(&xmltv_name, &xtream_name, epg_id_match, exact_name_match)
}

/// Detect changes in the provider's stream list by comparing with database.
#[tauri::command]
pub fn detect_provider_changes(
    db: State<DbConnection>,
    account_id: i32,
    current_streams: Vec<XtreamChannel>,
) -> Result<ProviderChanges, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::detect_provider_changes(&mut conn, account_id, &current_streams)
}

/// Auto-match new streams to XMLTV channels using fuzzy algorithm.
#[tauri::command]
pub fn auto_rematch_new_streams(
    db: State<DbConnection>,
    new_streams: Vec<XtreamChannel>,
    threshold: Option<f64>,
) -> Result<i32, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::auto_rematch_new_streams(&mut conn, &new_streams, threshold)
}

/// Handle removed streams by deleting auto-generated mappings and promoting backups.
#[tauri::command]
pub fn handle_removed_streams(
    db: State<DbConnection>,
    account_id: i32,
    removed_stream_ids: Vec<i32>,
) -> Result<(i32, i32), String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::handle_removed_streams(&mut conn, account_id, &removed_stream_ids)
}

/// Handle changed streams by updating metadata and recalculating match confidence.
#[tauri::command]
pub fn handle_changed_streams(
    db: State<DbConnection>,
    account_id: i32,
    changed_streams: Vec<ChangedStream>,
    threshold: Option<f64>,
) -> Result<i32, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::handle_changed_streams(&mut conn, account_id, &changed_streams, threshold)
}

/// Auto-match M3U channels to XMLTV channels using fuzzy matching.
#[tauri::command]
pub async fn auto_match_m3u_channels(
    app: AppHandle,
    db: State<'_, DbConnection>,
    source_id: Option<i32>,
    threshold: Option<f64>,
) -> Result<M3uAutoMatchResponse, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    let result = svc::auto_match_m3u_channels(&mut conn, source_id, threshold, |progress| {
        match &progress {
            MatchProgress::Starting { message } => {
                let _ = app.emit("m3u_match_progress", serde_json::json!({
                    "status": "starting", "message": message
                }));
            }
            MatchProgress::Saving { message } => {
                let _ = app.emit("m3u_match_progress", serde_json::json!({
                    "status": "saving", "message": message
                }));
            }
            MatchProgress::Complete { matched, unmatched } => {
                let _ = app.emit("m3u_match_progress", serde_json::json!({
                    "status": "complete", "matched": matched, "unmatched": unmatched
                }));
            }
        }
    })?;

    Ok(M3uAutoMatchResponse {
        success: true,
        matched_count: result.matched_count,
        unmatched_count: result.unmatched_count,
        total_m3u_channels: result.total_m3u_channels,
        total_xmltv_channels: result.total_xmltv_channels,
        duration_ms: result.duration_ms,
        mappings_created: result.mappings_created,
        message: format!(
            "Matched {} M3U channels to {} XMLTV channels. {} mappings created.",
            result.matched_count, result.total_xmltv_channels, result.mappings_created
        ),
    })
}

/// Get M3U auto-match results for display.
#[tauri::command]
pub fn get_m3u_auto_match_results(
    db: State<DbConnection>,
    source_id: Option<i32>,
) -> Result<Vec<M3uMatchResult>, String> {
    let mut conn = db
        .get_connection()
        .map_err(|e| format!("Database connection error: {}", e))?;

    svc::get_m3u_auto_match_results(&mut conn, source_id)
}

File: src/lib/api/matcher.ts
import { invoke } from './invoke';

// ============================================================================
// Channel Matching types and functions (Story 3-1)
// ============================================================================

/** Match type enum */
export type MatchType = 'exact_epg_id' | 'exact_name' | 'fuzzy' | 'none';

/** Match response type for run_channel_matching command */
export interface MatchResponse {
  success: boolean;
  matchedCount: number;
  unmatchedCount: number;
  totalXmltv: number;
  totalXtream: number;
  durationMs: number;
  message: string;
}

/** Match statistics type */
export interface MatchStats {
  totalXmltv: number;
  totalXtream: number;
  matched: number;
  unmatched: number;
  multipleMatches: number;
  durationMs: number;
}

/** Channel mapping type */
export interface ChannelMapping {
  id: number;
  xmltvChannelId: number;
  xtreamChannelId: number;
  matchConfidence: number | null;
  isManual: number | null;
  isPrimary: number | null;
  streamPriority: number | null;
  createdAt: string;
}

/** XMLTV channel settings type */
export interface XmltvChannelSettings {
  id: number;
  xmltvChannelId: number;
  isEnabled: number | null;
  plexDisplayOrder: number | null;
  createdAt: string;
  updatedAt: string;
}

/** Match progress event payload */
export interface MatchProgressEvent {
  status: 'starting' | 'saving' | 'complete';
  message?: string;
  matched?: number;
  unmatched?: number;
}

/**
 * Run the channel matching algorithm
 * Matches all XMLTV channels to Xtream streams using fuzzy matching.
 * @param threshold - Optional confidence threshold (0.0 to 1.0). Defaults to 0.85.
 * @returns Match response with statistics
 */
export async function runChannelMatching(threshold?: number): Promise<MatchResponse> {
  return invoke<MatchResponse>('run_channel_matching', { threshold });
}

/**
 * Get current match statistics from the database
 * @returns Match statistics
 */
export async function getMatchStats(): Promise<MatchStats> {
  return invoke<MatchStats>('get_match_stats');
}

/**
 * Get channel mappings for a specific XMLTV channel
 * @param xmltvChannelId - XMLTV channel ID
 * @returns List of channel mappings sorted by priority
 */
export async function getChannelMappingsForXmltv(xmltvChannelId: number): Promise<ChannelMapping[]> {
  return invoke<ChannelMapping[]>('get_channel_mappings_for_xmltv', { xmltvChannelId });
}

/**
 * Get XMLTV channel settings
 * @param xmltvChannelId - XMLTV channel ID
 * @returns Channel settings or null if not found
 */
export async function getXmltvChannelSettings(xmltvChannelId: number): Promise<XmltvChannelSettings | null> {
  return invoke<XmltvChannelSettings | null>('get_xmltv_channel_settings', { xmltvChannelId });
}

/**
 * Get the current matching threshold
 * @returns Current threshold value (0.0 to 1.0)
 */
export async function getMatchThreshold(): Promise<number> {
  const result = await invoke<number | { threshold: number }>('get_match_threshold');
  // REST API returns { threshold: number }, Tauri IPC returns bare number
  if (typeof result === 'object' && result !== null && 'threshold' in result) {
    return result.threshold;
  }
  return result as number;
}

/**
 * Set the matching threshold
 * @param threshold - New threshold value (0.0 to 1.0)
 */
export async function setMatchThreshold(threshold: number): Promise<void> {
  return invoke<void>('set_match_threshold', { threshold });
}

/**
 * Normalize a channel name (for testing/debugging)
 * @param name - Channel name to normalize
 * @returns Normalized channel name
 */
export async function normalizeChannelName(name: string): Promise<string> {
  return invoke<string>('normalize_channel_name', { name });
}

/**
 * Calculate match score between two channel names (for testing/debugging)
 * @param xmltvName - XMLTV channel name
 * @param xtreamName - Xtream stream name
 * @param epgIdMatch - Whether EPG IDs match
 * @param exactNameMatch - Whether normalized names match exactly
 * @returns Match score (0.0 to 1.0)
 */
export async function calculateMatchScore(
  xmltvName: string,
  xtreamName: string,
  epgIdMatch: boolean,
  exactNameMatch: boolean
): Promise<number> {
  return invoke<number>('calculate_match_score', {
    xmltvName,
    xtreamName,
    epgIdMatch,
    exactNameMatch,
  });
}

/**
 * Format confidence score as percentage
 * @param confidence - Confidence score (0.0 to 1.0)
 * @returns Formatted percentage string (e.g., "95%")
 */
export function formatConfidence(confidence: number | null): string {
  if (confidence === null) {
    return 'N/A';
  }
  return `${Math.round(confidence * 100)}%`;
}

/**
 * Get match type display name
 * @param matchType - Match type
 * @returns Human-readable match type string
 */
export function getMatchTypeDisplay(matchType: MatchType): string {
  switch (matchType) {
    case 'exact_epg_id':
      return 'EPG ID Match';
    case 'exact_name':
      return 'Exact Name';
    case 'fuzzy':
      return 'Fuzzy Match';
    case 'none':
      return 'No Match';
    default:
      return 'Unknown';
  }
}

// ============================================================================
// M3U Auto-Match (Multi-Source Stream Support)
// ============================================================================

/** M3U auto-match response type */
export interface M3uAutoMatchResponse {
  success: boolean;
  matchedCount: number;
  unmatchedCount: number;
  totalM3uChannels: number;
  totalXmltvChannels: number;
  durationMs: number;
  mappingsCreated: number;
  message: string;
}

/** M3U match result type */
export interface M3uMatchResult {
  xmltvChannelId: number;
  m3uChannelId: number;
  confidence: number;
  isPrimary: boolean;
  streamPriority: number;
  matchType: MatchType;
}

/**
 * Auto-match M3U channels to XMLTV channels using fuzzy matching
 * @param sourceId - Optional M3U source ID to match. If not provided, matches all sources.
 * @param threshold - Optional confidence threshold (0.0 to 1.0). Defaults to 0.85.
 * @returns Match response with statistics
 */
export async function autoMatchM3uChannels(
  sourceId?: number,
  threshold?: number
): Promise<M3uAutoMatchResponse> {
  return invoke<M3uAutoMatchResponse>('auto_match_m3u_channels', {
    sourceId,
    threshold,
  });
}

/**
 * Get M3U auto-match results
 * @param sourceId - Optional M3U source ID to filter by
 * @returns List of M3U match results
 */
export async function getM3uAutoMatchResults(sourceId?: number): Promise<M3uMatchResult[]> {
  return invoke<M3uMatchResult[]>('get_m3u_auto_match_results', { sourceId });
}

// ============================================================================
// XMLTV Channel Display types and functions (Story 3-2)
// ============================================================================

/** Matched Xtream stream info for display */
export interface XtreamStreamMatch {
  id: number;
  mappingId: number;
  name: string;
  streamIcon: string | null;
  qualities: string[];
  matchConfidence: number;
  isPrimary: boolean;
  isManual: boolean;
  streamPriority: number;
  /** True if this is a manual match pointing to a stream that no longer exists */
  isOrphaned: boolean;
}

/** XMLTV channel with all mapping info for display */
export interface XmltvChannelWithMappings {
  id: number;
  sourceId: number;
  channelId: string;
  displayName: string;
  icon: string | null;
  isSynthetic: boolean;
  // Settings
  isEnabled: boolean;
  plexDisplayOrder: number | null;
  // Matches
  matchCount: number;
  matches: XtreamStreamMatch[];
  // Tags
  tags: string[];
}

/**
 * Get all XMLTV channels with their mapped Xtream streams
 * @returns List of XMLTV channels with mapping info
 */
export async function getXmltvChannelsWithMappings(): Promise<XmltvChannelWithMappings[]> {
  return invoke<XmltvChannelWithMappings[]>('get_xmltv_channels_with_mappings');
}

/**
 * Set the primary stream for an XMLTV channel
 * @param xmltvChannelId - XMLTV channel ID
 * @param xtreamChannelId - Xtream stream ID to make primary
 * @returns Updated list of matches
 */
export async function setPrimaryStream(
  xmltvChannelId: number,
  xtreamChannelId: number
): Promise<XtreamStreamMatch[]> {
  return invoke<XtreamStreamMatch[]>('set_primary_stream', {
    xmltvChannelId,
    xtreamChannelId,
  });
}

/**
 * Toggle the enabled status of an XMLTV channel
 * @param channelId - XMLTV channel ID
 * @returns Updated channel with mappings
 */
export async function toggleXmltvChannel(
  channelId: number
): Promise<XmltvChannelWithMappings> {
  return invoke<XmltvChannelWithMappings>('toggle_xmltv_channel', { channelId });
}

/**
 * Get match count label for display
 * @param count - Number of matched streams
 * @returns Formatted string (e.g., "1 stream", "3 streams")
 */
export function getMatchCountLabel(count: number): string {
  if (count === 0) {
    return 'No stream matched';
  }
  return `${count} stream${count === 1 ? '' : 's'}`;
}

/**
 * Get quality badge classes for styling
 * @param quality - Quality tier (SD, HD, FHD, 4K)
 * @returns Tailwind CSS classes for the badge
 */
export function getQualityBadgeClasses(quality: string): string {
  switch (quality.toUpperCase()) {
    case '4K':
      return 'bg-purple-100 text-purple-800';
    case 'FHD':
      return 'bg-blue-100 text-blue-800';
    case 'HD':
      return 'bg-green-100 text-green-800';
    case 'SD':
    default:
      return 'bg-gray-100 text-gray-800';
  }
}

// ============================================================================
// Manual Stream Matching types and functions (Story 3-3)
// ============================================================================

/** Xtream stream for search dropdown */
export interface XtreamStreamSearchResult {
  id: number;
  streamId: number;
  name: string;
  streamIcon: string | null;
  qualities: string[];
  categoryName: string | null;
  /** List of XMLTV channel IDs this stream is already matched to */
  matchedToXmltvIds: number[];
  /** Fuzzy match score against search query (0.0-1.0), null if no search query */
  fuzzyScore: number | null;
}

/**
 * Get all Xtream streams for the search dropdown
 * @returns List of Xtream streams with their current mappings
 */
export async function getAllXtreamStreams(): Promise<XtreamStreamSearchResult[]> {
  return invoke<XtreamStreamSearchResult[]>('get_all_xtream_streams');
}

/**
 * Search Xtream streams by fuzzy matching against a query string
 * @param query - Search query (e.g., XMLTV channel name)
 * @returns List of Xtream streams with fuzzy scores, ordered by score descending
 */
export async function searchXtreamStreams(query: string): Promise<XtreamStreamSearchResult[]> {
  return invoke<XtreamStreamSearchResult[]>('search_xtream_streams', { query });
}

/**
 * Add a manual stream mapping between an XMLTV channel and an Xtream stream
 * @param xmltvChannelId - XMLTV channel ID
 * @param xtreamChannelId - Xtream stream ID to map
 * @param setAsPrimary - Whether to set this stream as primary
 * @returns Updated list of matches for the XMLTV channel
 */
export async function addManualStreamMapping(
  xmltvChannelId: number,
  xtreamChannelId: number,
  setAsPrimary: boolean
): Promise<XtreamStreamMatch[]> {
  return invoke<XtreamStreamMatch[]>('add_manual_stream_mapping', {
    xmltvChannelId,
    xtreamChannelId,
    setAsPrimary,
  });
}

/**
 * Remove a stream mapping
 * @param mappingId - Mapping ID to remove
 * @returns Updated list of matches for the XMLTV channel
 */
export async function removeStreamMapping(
  mappingId: number
): Promise<XtreamStreamMatch[]> {
  return invoke<XtreamStreamMatch[]>('remove_stream_mapping', { mappingId });
}
